Avoid some undefined behavior in test_bruteforce.

hvdijk opened the pull request:
* Ulp_Error*: ilogb(reference) - 1 may overflow if reference is zero.
* binary_i_double Test: DoubleFromUInt32's result is a cl_double and the attempt is to store it as a cl_double, but p was defined as a pointer to cl_ulong, resulting in an unintended implicit conversion that is not valid for out-of-range doubles.
* exp2, tanpi: ensure early exit for NaN.
* shift_right_sticky_128: avoid out-of-range shift if shift value is exactly 64.
* scalbn: e += n may overflow if n is large, move it after the check for large n.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/OpenCL-CTS?pullRequest=2400) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/OpenCL-CTS?pullRequest=2400) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/OpenCL-CTS?pullRequest=2400) it.</sub>

hvdijk commented:
Background: after updating our build infrastructure to build OpenCL CTS with a newer compiler version, I got this result:
```
ldexp...
144:           ldexp fp32     .
ERROR: ldexp: inf ulp error at {0x1.6694eep+20 (0x49b34a77), 2147483207}: *0x0p+0 (0x00000000) vs. inf (0x7f800000) at index: 12055
```
That is, our implementation was correctly returning infinity for `ldexp(0x1.6694eep+20, 2147483207)`, but the OpenCL CTS reference result was computed to be zero. When seeing that, I built OpenCL CTS with `clang -fsanitize=undefined` and addressed the issues that came up when running `test_brute_force`.

svenvh reviewed (approved):
LGTM, thanks for the fixes!

lakshmih commented:
Reviewing, could you hold on till end of week. Thanks

bashbaug commented:
Merging as discussed in the July 8th teleconference.
